include/raft.h

```diff
@@ -201,74 +205,34 @@ struct raft_entry
- * servers) its refcount is increased by one. Whenever an entry gets deleted
- * from the log its refcount is decreased by one. Likewise, whenever an I/O
- * request is completed the refcount of the relevant entries is decreased by
- * one. When the refcount drops to zero the memory that its @buf attribute
- * points to gets released, or, if the @batch attribute is non-NULL, a check is
- * made to see if all other entries of the same batch also have a zero refcount,
- * and the memory that @batch points to gets released if that's the case.
- */
-struct raft_entry_ref
-{
-    raft_term term;              /* Term of the entry being ref-counted. */
-    raft_index index;            /* Index of the entry being ref-counted. */
-    unsigned short count;        /* Number of references. */
-    struct raft_entry_ref *next; /* Next item in the bucket (for collisions). */
-};
-
-/**
- * In-memory cache of the persistent raft log stored on disk.
- *
- * The raft log cache is implemented as a circular buffer of log entries, which
- * makes some frequent operations very efficient (e.g. deleting the first N
- * entries when snapshotting).
- */
-struct raft_log
-{
-    struct raft_entry *entries;  /* Circular buffer of log entries. */
-    size_t size;                 /* Number of available slots in the buffer. */
-    size_t front, back;          /* Indexes of used slots [front, back). */
-    raft_index offset;           /* Index of first entry is offset+1. */
-    struct raft_entry_ref *refs; /* Log entries reference counts hash table. */
-    size_t refs_size;            /* Size of the reference counts hash table. */
-    struct                       /* Information about last snapshot, or zero. */
-    {
-        raft_index last_index; /* Snapshot replaces all entries up to here. */
-        raft_term last_term;   /* Term of last index. */
-    } snapshot;
-};
-
 /**
  * Hold the arguments of a RequestVote RPC.
  *
  * The RequestVote RPC is invoked by candidates to gather votes.
  */
 struct raft_request_vote
 {
+    int version;
     raft_term term;            /* Candidate's term. */
     raft_id candidate_id;      /* ID of the server requesting the vote. */
     raft_index last_log_index; /* Index of candidate's last log entry. */
     raft_index last_log_term;  /* Term of log entry at last_log_index. */
     bool disrupt_leader;       /* True if current leader should be discarded. */
     bool pre_vote;             /* True if this is a pre-vote request. */
 };
+#define RAFT_REQUEST_VOTE_VERSION 2
 
 /**
  * Hold the result of a RequestVote RPC.
  */
 struct raft_request_vote_result
 {
+    int version;
     raft_term term;        /* Receiver's current term (candidate updates itself). */
     bool vote_granted;     /* True means candidate received vote. */
-    raft_tribool pre_vote; /* The response to a pre-vote RequestVote or not. */
+    bool pre_vote;         /* The response to a pre-vote RequestVote or not. */
```

Comment:
This is a nice change, if I understand correctly the `raft_tribool` type is not needed anymore because we version messages? However, won't there issues with backward compatibility?

I didn't quite look in detail, just asking to conform it's ok.

Reply:
Backwards compatibility on the wire you mean?

Reply:
> Backwards compatibility on the wire you mean?

Yes, I believe `0` currently has a special meaning (i.e. neither true or false, iow, the other side didn't include it). If we move to regular `bool` that will be `false`. As said, I'm not entirely sure of the details, I was wondering if problems could arise from an old node talking to a new node.

In any case, all things considered, even if there were issue, I think they would be minor and unlikely enough to not bother and keep this change.

Reply:
It should be okay, it's just a different translation to the new struct of a flags argument of `RequestVoteResult` on the wire. In the past if we received the old message we would fill out the struct with `raft_tribool_unknown`, now we fill it out with `version = 1` and ignore the new bool `pre_vote` struct field.